x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -172,4 +210,49 @@ private void removeSessionForShard(String sessionUUID, IndexShard indexShard) {
             }
         }
     }
+
+    private static class RefCountedCloseable<T extends Closeable> extends AbstractRefCounted {
+
+        private final String name;
+        private final T object;
+
+        private RefCountedCloseable(String name, T object) {
+            super(name);
+            this.name = name;
+            this.object = object;
+        }
+
+        @Override
+        protected void closeInternal() {
+            IOUtils.closeWhileHandlingException(object);
+        }
+    }
+
+    public static class Reader implements Closeable {
+
+        private final RefCountedCloseable<Engine.IndexCommitRef> session;
+        private final RefCountedCloseable<IndexInput> input;
+
+        private Reader(RefCountedCloseable<Engine.IndexCommitRef> session, RefCountedCloseable<IndexInput> input) {
+            this.session = session;
+            this.input = input;
+            session.incRef();
+            input.incRef();
```

Comment:
we need to make sure we release the session if this one fails?

Reply:
I added some releases inside the ctor in case something fails.